# HG changeset patch
# User Torsten <mttl@mailbox.org>
# Date 1543352665 -3600
#      Tue Nov 27 22:04:25 2018 +0100
# Node ID 494452db0477638177dc6805beac3651cd668355
# Parent  c7f845bf4fec3e7258c1320ef51b06c122f2c181
doc browser: append anchor from url to history titles (bug #54938)

* documentation.cc (update_history): extract anchor from url, replace
  encodings by normal characters and append it to the title before
  adding the title to the history menus

diff --git a/libgui/src/documentation.cc b/libgui/src/documentation.cc
--- a/libgui/src/documentation.cc
+++ b/libgui/src/documentation.cc
@@ -716,16 +716,45 @@ namespace octave
     int count = qMin (new_count, int (max_history_entries));
 
     // Fill used menu entries
     for (int i = 0; i < count; i++)
       {
         QString title = m_doc_browser->historyTitle (prev_next*(i+1));
         title.remove (QRegExp ("\\s*\\(*GNU Octave \\(version [^\\)]*\\)[: \\)]*"));
 
+        // Sinve the title only contains the section name and not the
+        // specific anchor, extract the latter from the url and append
+        // it to the title
+        QString url = m_doc_browser->historyUrl (prev_next*(i+1)).toString (QUrl::FullyEncoded);
+        if (url.contains ('#'))
+          {
+            // Get the anchor from the url
+            QString anchor = url.split ('#').last ();
+
+            // Remove internal string parts
+            anchor.remove (QRegExp ("^index-"));
+            anchor.remove (QRegExp ("^SEC_"));
+            anchor.remove (QRegExp ("^XREF"));
+            anchor.remove ("Concept-Index_cp_letter-");
+            anchor.replace ("-"," ");
+
+            // replace encoded special chars by there unencoded versions
+            QRegExp rx = QRegExp ("_00([0-7][0-9a-f])");
+            int pos = 0;
+            while ((pos = rx.indexIn(anchor, pos)) != -1)
+              {
+                anchor.replace ("_00"+rx.cap (1), QChar (rx.cap (1).toInt (nullptr,16)));
+                pos += rx.matchedLength();
+              }
+
+            if (title != anchor)
+              title = title + ": " + anchor;
+          }
+
         if (i == 0)
           a->setText (title); // set tool tip for prev/next buttons
 
         actions[i]->setText (title);
         actions[i]->setData (m_doc_browser->historyUrl (prev_next*(i+1)));
         actions[i]->setEnabled (true);
         actions[i]->setVisible (true);
       }
